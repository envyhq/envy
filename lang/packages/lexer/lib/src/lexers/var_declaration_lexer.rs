use super::{attribute_block_lexer::AttributeBlockLexer, utils::lookbehind_raw_token};
use crate::{
    chars::LexerChar,
    lexers::utils::{is_newline, is_whitespace},
    tokens::{LexerSymbol, LexerToken, LexerTokenKind, LexerType, TokenPosition},
    Lexer,
};
use strum::IntoEnumIterator;

pub struct VarDeclarationLexer<'a> {
    pub tokens: Vec<LexerToken>,

    chars: &'a [String],
    buffer: Vec<String>,

    start_position: TokenPosition,
}

impl<'a> VarDeclarationLexer<'a> {
    pub fn new(chars: &'a [String], start_position: TokenPosition) -> Self {
        Self {
            chars,
            tokens: vec![],
            buffer: vec![],
            start_position,
        }
    }

    fn buffer_to_type(&self, buffer: Vec<String>) -> Option<LexerType> {
        let buffered = buffer.join("");

        LexerType::iter().find(|type_value| type_value.to_string() == buffered)
    }

    fn lex_newline(&mut self, current_position: &mut TokenPosition) {
        self.lex_type_value(current_position, Some(LexerChar::NewLine));

        self.tokens.push(LexerToken::new(
            LexerTokenKind::Symbol(LexerSymbol::Newline),
            current_position.clone(),
            current_position.clone(),
        ));

        current_position.line += 1;
        current_position.column = 0;
    }

    fn lex_whitespace(&mut self, current_position: &mut TokenPosition) {
        self.lex_type_value(current_position, None);
        self.buffer.clear();
    }

    fn lex_block_close(&mut self, current_position: &mut TokenPosition) {
        self.tokens.push(LexerToken::new(
            LexerTokenKind::Symbol(LexerSymbol::BlockCloseCurly),
            current_position.clone(),
            current_position.clone(),
        ));
    }

    fn lex_var_assignment(&mut self, current_position: &mut TokenPosition) {
        let (buffer, from, to) = lookbehind_raw_token(
            current_position,
            &self.buffer,
            Some(LexerChar::VarAssignmentColon),
        );

        let buffered = buffer.join("");
        if !buffered.is_empty() {
            self.tokens.push(LexerToken::new(
                LexerTokenKind::Identifier(buffered.clone()),
                from,
                to,
            ));
        } else {
            panic!("Expected variable declaration identifier before colon")
        }
        self.tokens.push(LexerToken::new(
            LexerTokenKind::Symbol(LexerSymbol::VarAssignmentColon),
            current_position.clone(),
            current_position.clone(),
        ));
        self.buffer.clear();
    }

    fn lex_block_open(
        &mut self,
        current_position: &mut TokenPosition,
        processed_count: &mut usize,
        index: &usize,
    ) {
        self.lex_type_value(current_position, Some(LexerChar::BlockOpenCurly));

        self.tokens.push(LexerToken::new(
            LexerTokenKind::Symbol(LexerSymbol::BlockOpenCurly),
            current_position.clone(),
            current_position.clone(),
        ));

        let sub_chars = &self.chars[(index + 1)..];

        let mut block_lexer = AttributeBlockLexer::new(sub_chars, current_position.clone());

        let (block_count, block_position) = block_lexer.lex();

        *current_position = block_position;

        *processed_count += block_count;

        self.tokens.append(&mut block_lexer.tokens);
    }

    fn lex_type_value(
        &mut self,
        current_position: &mut TokenPosition,
        stop_char: Option<LexerChar>,
    ) {
        let (buffer, from, to) = lookbehind_raw_token(current_position, &self.buffer, stop_char);
        let type_value = self.buffer_to_type(buffer);
        if let Some(type_value) = type_value {
            self.tokens.push(LexerToken::new(
                LexerTokenKind::Type(type_value.clone()),
                from,
                to,
            ));
        }
    }
}

impl<'a> Lexer for VarDeclarationLexer<'a> {
    fn lex(&mut self) -> (usize, TokenPosition) {
        let chars = self.chars.iter().enumerate();

        let mut processed_count = 0;
        let mut current_position = self.start_position.clone();

        for (index, char) in chars {
            let char = char.to_owned();

            processed_count += 1;
            current_position.column += 1;

            self.buffer.push(char.clone());

            if is_newline(&char) {
                self.lex_newline(&mut current_position);
                return (processed_count, current_position);
            }

            if is_whitespace(&char) {
                self.lex_whitespace(&mut current_position);
                continue;
            }

            match char.parse() {
                Ok(LexerChar::BlockCloseCurly) => {
                    self.lex_block_close(&mut current_position);
                    return (processed_count, current_position);
                }
                Ok(LexerChar::VarAssignmentColon) => {
                    self.lex_var_assignment(&mut current_position);
                    continue;
                }
                Ok(LexerChar::BlockOpenCurly) => {
                    self.lex_block_open(&mut current_position, &mut processed_count, &index);
                    return (processed_count, current_position);
                }
                _ => {
                    continue;
                }
            }
        }

        self.lex_type_value(&mut current_position, None);

        (processed_count, current_position)
    }
}

#[cfg(test)]
mod tests {
    use crate::{lexers::var_declaration_lexer::VarDeclarationLexer, Lexer, TokenPosition};
    use envy_unit_testing::str_to_graphemes;

    #[test]
    fn lexes_var_tokens() {
        // We dont have "var" keyword here, its handled by the source file lexer
        let input = str_to_graphemes("id: int");

        let start_line = 11;
        let start_column = 40;
        let mut lexer =
            VarDeclarationLexer::new(&input, TokenPosition::new(start_line, start_column));

        let (count, position) = lexer.lex();

        assert_eq!(count, input.len());
        assert_eq!(position, TokenPosition::new(start_line, start_column + 7));
        assert_eq!(lexer.tokens.len(), 3);
        insta::assert_yaml_snapshot!(lexer.tokens);
    }
}
