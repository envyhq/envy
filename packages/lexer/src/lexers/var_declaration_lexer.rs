use super::{attribute_block_lexer::AttributeBlockLexer, utils::lookbehind_raw_token};
use crate::{
    chars::LexerChar,
    lexers::utils::{is_newline, is_whitespace},
    tokens::{LexerSymbol, LexerToken, LexerTokenKind, LexerType, TokenPosition},
    Lexer,
};
use std::str::FromStr;
use strum::IntoEnumIterator;

pub struct VarDeclarationLexer<'a> {
    pub tokens: Vec<LexerToken>,

    chars: &'a [String],
    buffer: Vec<String>,

    start_position: TokenPosition,
}

impl<'a> VarDeclarationLexer<'a> {
    pub fn new(chars: &'a [String], start_position: TokenPosition) -> Self {
        Self {
            chars,
            tokens: vec![],
            buffer: vec![],
            start_position,
        }
    }

    fn buffer_to_type(&self) -> Option<LexerType> {
        let buffered = self.buffer.join("");

        LexerType::iter().find(|type_value| type_value.to_string() == buffered)
    }
}

impl<'a> Lexer for VarDeclarationLexer<'a> {
    // Called whenever the lexer encounters a var keyword, we continue to lex in the context of variable declaration, having already stored the var keyword token.
    // We search for a colon to indicate the start of the variable type assignment, taking everything before that as the identifier and everything after as the type.
    // We are given the whole source file from the var keyword onwards, so we lex until we reach a valid output.
    // The source file lexer must continue where we left of, so we return the number of characters we have processed.
    fn lex(&mut self) -> (usize, TokenPosition) {
        let chars = self.chars.iter().enumerate();

        self.buffer.clear();

        let mut processed_count = 0;
        let mut current_position = self.start_position.clone();

        for (index, char) in chars {
            let char = char.to_owned();

            processed_count += 1;
            current_position.column += 1;

            self.buffer.push(char.clone());

            if is_newline(&char) {
                self.buffer.pop(); //  remove the newline
                let type_value = self.buffer_to_type();
                if let Some(type_value) = type_value {
                    self.tokens.push(LexerToken::new(
                        LexerTokenKind::Type(type_value.clone()),
                        TokenPosition::new(
                            current_position.line,
                            current_position.column - type_value.to_string().len(),
                        ),
                        TokenPosition::new(
                            current_position.line,
                            current_position.column - 1, // -1 because buffer.pop() to remove newline
                        ),
                    ));
                }

                self.tokens.push(LexerToken::new(
                    LexerTokenKind::Symbol(LexerSymbol::Newline),
                    current_position.clone(),
                    current_position.clone(),
                ));

                current_position.line += 1;
                current_position.column = 0;

                return (processed_count, current_position);
            }

            if is_whitespace(&char) {
                self.buffer.pop(); //  remove the whitespace
                let type_value = self.buffer_to_type();
                if let Some(type_value) = type_value {
                    self.tokens.push(LexerToken::new(
                        LexerTokenKind::Type(type_value.clone()),
                        TokenPosition::new(
                            current_position.line,
                            current_position.column - type_value.to_string().len(),
                        ),
                        TokenPosition::new(
                            current_position.line,
                            current_position.column - 1, // -1 because buffer.pop() to remove newline
                        ),
                    ));
                }

                self.buffer.clear();
                continue;
            }

            match LexerChar::from_str(&char) {
                Ok(LexerChar::BlockCloseCurly) => {
                    // Terminate lexing if we reach a block close curly, return
                    self.tokens.push(LexerToken::new(
                        LexerTokenKind::Symbol(LexerSymbol::BlockCloseCurly),
                        current_position.clone(),
                        current_position.clone(),
                    ));

                    return (processed_count, current_position);
                }
                Ok(LexerChar::VarAssignmentColon) => {
                    // When we reach the colon, lex the var idenitifier and push it to the tokens before the colon
                    let (buffer, from, to) = lookbehind_raw_token(
                        &current_position,
                        &self.buffer,
                        Some(LexerChar::VarAssignmentColon),
                    );

                    let buffered = buffer.join("");
                    if !buffered.is_empty() {
                        self.tokens.push(LexerToken::new(
                            LexerTokenKind::Identifier(buffered.clone()),
                            from,
                            to,
                        ));
                    } else {
                        panic!("Expected variable declaration identifier before colon")
                    }
                    self.tokens.push(LexerToken::new(
                        LexerTokenKind::Symbol(LexerSymbol::VarAssignmentColon),
                        current_position.clone(),
                        current_position.clone(),
                    ));
                    self.buffer.clear();
                    continue;
                }
                Ok(LexerChar::BlockOpenCurly) => {
                    // If we reach a block open curly, check current buffer for type and continue to lex for attributes
                    let type_value = self.buffer_to_type();
                    if let Some(type_value) = type_value {
                        self.tokens.push(LexerToken::new(
                            LexerTokenKind::Type(type_value.clone()),
                            TokenPosition::new(
                                current_position.line,
                                current_position.column - type_value.to_string().len(),
                            ),
                            current_position.clone(),
                        ));
                    }

                    self.tokens.push(LexerToken::new(
                        LexerTokenKind::Symbol(LexerSymbol::BlockOpenCurly),
                        current_position.clone(),
                        current_position.clone(),
                    ));

                    let sub_chars = &self.chars[(index + 1)..];

                    let mut block_lexer =
                        AttributeBlockLexer::new(sub_chars, current_position.clone());

                    let (block_count, block_position) = block_lexer.lex();

                    current_position = block_position;

                    processed_count += block_count;

                    self.tokens.append(&mut block_lexer.tokens);

                    return (processed_count, current_position);
                }
                _ => {
                    continue;
                }
            }
        }

        // If we haven't returned already, we reached the end of the input. Check the buffer for any remaining token.
        let type_value = self.buffer_to_type();
        if let Some(type_value) = type_value {
            self.tokens.push(LexerToken::new(
                LexerTokenKind::Type(type_value.clone()),
                TokenPosition::new(
                    current_position.line,
                    current_position.column + 1 - type_value.to_string().len(),
                ),
                current_position.clone(),
            ));
        }

        (processed_count, current_position)
    }
}

#[cfg(test)]
mod tests {
    use crate::{lexers::var_declaration_lexer::VarDeclarationLexer, Lexer, TokenPosition};
    use nv_unit_testing::str_to_graphemes;

    #[test]
    fn lexes_var_tokens() {
        // We dont have "var" keyword here, its handled by the source file lexer
        let input = str_to_graphemes("id: int");

        let start_line = 11;
        let start_column = 40;
        let mut lexer =
            VarDeclarationLexer::new(&input, TokenPosition::new(start_line, start_column));

        let (count, position) = lexer.lex();

        assert_eq!(count, input.len());
        assert_eq!(position, TokenPosition::new(start_line, start_column + 7));
        assert_eq!(lexer.tokens.len(), 3);
        insta::assert_yaml_snapshot!(lexer.tokens);
    }
}
