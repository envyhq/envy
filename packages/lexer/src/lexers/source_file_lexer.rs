use super::var_declaration_lexer::VarDeclarationLexer;
use crate::{
    buffers::identifiers::buffer_to_keyword,
    lexer::LexerResult,
    lexers::{
        module_declaration_lexer::ModuleDeclarationLexer,
        provider_declaration_lexer::ProviderDeclarationLexer,
        utils::{is_newline, is_whitespace},
    },
    tokens::{
        LexerDeclarationKeyword, LexerKeyword, LexerSymbol, LexerToken, LexerTokenKind,
        TokenPosition,
    },
    Lexer,
};
use unicode_segmentation::UnicodeSegmentation;

pub struct SourceFileLexer {
    pub tokens: Vec<LexerToken>,

    chars: Vec<String>,
    buffer: Vec<String>,
}

impl SourceFileLexer {
    pub fn new(input: &str) -> Self {
        let chars = input
            .graphemes(true)
            .map(|char| char.to_owned())
            .collect::<Vec<String>>();

        Self {
            chars,
            tokens: vec![],
            buffer: vec![],
        }
    }
    fn buffer_to_keyword(&self) -> Option<LexerKeyword> {
        buffer_to_keyword(&self.buffer)
    }
}

impl Lexer for SourceFileLexer {
    fn lex(&mut self) -> (usize, TokenPosition) {
        let mut chars_iter = self.chars.iter().enumerate();

        if !self.tokens.is_empty() || !self.buffer.is_empty() {
            self.tokens = vec![];
            self.buffer = vec![];
        }

        let mut processed_count = 0;
        let mut current_position = TokenPosition::new(1, 0);

        while let Some((index, char)) = chars_iter.next() {
            let char = char.to_owned();

            processed_count += 1;
            current_position.column += 1;

            if is_newline(&char) {
                self.tokens.push(LexerToken::new(
                    LexerTokenKind::Symbol(LexerSymbol::Newline),
                    current_position.clone(),
                    current_position.clone(),
                ));
                current_position.line += 1;
                current_position.column = 0;

                continue;
            }

            if is_whitespace(&char) {
                continue;
            }

            self.buffer.push(char.clone());

            if let Some(token) = self.buffer_to_keyword() {
                self.tokens.push(LexerToken::new(
                    LexerTokenKind::Keyword(token.clone()),
                    TokenPosition {
                        line: current_position.line,
                        column: current_position
                            .column
                            .saturating_sub(token.to_string().len() - 1),
                    },
                    current_position.clone(),
                ));
                self.buffer.clear();
                let sub_chars = &self.chars[(index + 1)..];

                let result: LexerResult = match token {
                    LexerKeyword::DeclarationKeyword(LexerDeclarationKeyword::Var) => {
                        let mut lexer =
                            VarDeclarationLexer::new(sub_chars, current_position.clone());
                        let (count, final_position) = lexer.lex();

                        LexerResult {
                            processed_count: count,
                            tokens: lexer.tokens,
                            final_position,
                        }
                    }
                    LexerKeyword::DeclarationKeyword(LexerDeclarationKeyword::Module) => {
                        let mut lexer =
                            ModuleDeclarationLexer::new(sub_chars, current_position.clone());
                        let (count, final_position) = lexer.lex();

                        LexerResult {
                            processed_count: count,
                            tokens: lexer.tokens,
                            final_position,
                        }
                    }
                    LexerKeyword::DeclarationKeyword(LexerDeclarationKeyword::Provider) => {
                        let mut lexer =
                            ProviderDeclarationLexer::new(sub_chars, current_position.clone());
                        let (count, final_position) = lexer.lex();

                        LexerResult {
                            processed_count: count,
                            tokens: lexer.tokens,
                            final_position,
                        }
                    }
                    _ => {
                        continue;
                    }
                };

                self.tokens.append(&mut result.tokens.clone());

                processed_count += result.processed_count;
                if result.processed_count > 0 {
                    chars_iter.nth(result.processed_count - 1);
                }
                current_position = result.final_position;
                self.buffer.clear();

                continue;
            }
        }

        (processed_count, current_position)
    }
}

#[cfg(test)]
mod tests {
    use crate::{lexers::source_file_lexer::SourceFileLexer, Lexer, TokenPosition};

    #[test]
    fn lexes_source_file_tokens() {
        let input = "var my_test_var: int";

        let mut lexer = SourceFileLexer::new(input);

        let (count, position) = lexer.lex();

        assert_eq!(count, input.len());
        assert_eq!(position, TokenPosition::new(1, 20));
        assert_eq!(lexer.tokens.len(), 4);

        insta::assert_yaml_snapshot!(lexer.tokens);
    }

    #[test]
    fn lexes_source_file_tokens_with_pub_var() {
        let input = "var my_test_var: int

pub var my_pub_test_var: str";

        let mut lexer = SourceFileLexer::new(input);

        let (count, position) = lexer.lex();

        assert_eq!(count, input.len());
        assert_eq!(position, TokenPosition::new(3, 28));
        assert_eq!(lexer.tokens.len(), 11);

        insta::assert_yaml_snapshot!(lexer.tokens);
    }

    #[test]
    fn lexes_source_file_tokens_with_var_attributes() {
        let input = "var my_test_var: int

pub var my_pub_test_var: str
var my_attr_test_var: url {
    protocol = \"https\"
    host = \"test.com\"
}";

        let mut lexer = SourceFileLexer::new(input);

        let (count, position) = lexer.lex();

        assert_eq!(count, input.len());
        assert_eq!(position, TokenPosition::new(7, 1));
        assert_eq!(lexer.tokens.len(), 27);

        insta::assert_yaml_snapshot!(lexer.tokens);
    }

    #[test]
    fn lexes_source_file_tokens_with_module() {
        let input = "var my_test_var: int

module TestModule {
    var my_test_var: int
}";

        let mut lexer = SourceFileLexer::new(input);

        let (count, position) = lexer.lex();

        assert_eq!(count, input.len());
        assert_eq!(position, TokenPosition::new(5, 1));
        assert_eq!(lexer.tokens.len(), 16);

        insta::assert_yaml_snapshot!(lexer.tokens);
    }

    #[test]
    fn lexes_source_file_tokens_with_provider() {
        let input = "var my_test_var: int

provider Env: env

module TestModule {
    var my_test_var: int
}";

        let mut lexer = SourceFileLexer::new(input);

        let (count, position) = lexer.lex();

        assert_eq!(count, input.len());
        assert_eq!(position, TokenPosition::new(7, 1));
        assert_eq!(lexer.tokens.len(), 22);

        insta::assert_yaml_snapshot!(lexer.tokens);
    }
}
