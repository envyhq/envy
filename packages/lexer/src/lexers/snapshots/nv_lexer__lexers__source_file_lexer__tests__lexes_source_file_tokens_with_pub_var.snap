---
source: packages/lexer/src/lexers/source_file_lexer.rs
expression: lexer.tokens
---
- kind:
    Keyword:
      DeclarationKeyword: Var
  range:
    from:
      line: 1
      column: 1
    to:
      line: 1
      column: 3
- kind:
    Identifier: my_test_var
  range:
    from:
      line: 1
      column: 5
    to:
      line: 1
      column: 15
- kind:
    Symbol: VarAssignmentColon
  range:
    from:
      line: 1
      column: 16
    to:
      line: 1
      column: 16
- kind:
    Type: Int
  range:
    from:
      line: 1
      column: 18
    to:
      line: 1
      column: 20
- kind:
    Symbol: Newline
  range:
    from:
      line: 1
      column: 21
    to:
      line: 1
      column: 21
- kind:
    Symbol: Newline
  range:
    from:
      line: 2
      column: 1
    to:
      line: 2
      column: 1
- kind:
    Keyword:
      VarModifierKeyword: Pub
  range:
    from:
      line: 3
      column: 1
    to:
      line: 3
      column: 3
- kind:
    Keyword:
      DeclarationKeyword: Var
  range:
    from:
      line: 3
      column: 5
    to:
      line: 3
      column: 7
- kind:
    Identifier: my_pub_test_var
  range:
    from:
      line: 3
      column: 9
    to:
      line: 3
      column: 23
- kind:
    Symbol: VarAssignmentColon
  range:
    from:
      line: 3
      column: 24
    to:
      line: 3
      column: 24
- kind:
    Type: String
  range:
    from:
      line: 3
      column: 26
    to:
      line: 3
      column: 28
